### PR TITLE
Add retry logic for failed watches

### LINKED ISSUE #2: HA Controller restarting too many times
After the piraeus-operator 1.3.0 release (https://github.com/piraeusdatastore/piraeus-operator/releases/tag/v1.3.0), I enabled the HA Controller in Helm as described. Since then the HA Controller has been restarted 150 times, which is seemingly high. A Nagios monitoring script is alerting after some 10s of pod restarts, so this alert would always fire after some time even if the pod is deleted to reset the restart counter. Please look into it if it is a bug and is fixable. 

```bash
$ k get po
NAME                                         READY   STATUS    RESTARTS   AGE
piraeus-op-cs-controller-5db495d656-gnkv5    1/1     Running   4          5d
piraeus-op-csi-controller-6ccd9fbc44-cgw49   6/6     Running   0          3d22h
piraeus-op-csi-node-5j2f2                    3/3     Running   0          3d22h
piraeus-op-csi-node-7rbsr                    3/3     Running   0          3d22h
piraeus-op-csi-node-hdqhx                    3/3     Running   0          3d22h
piraeus-op-etcd-0                            1/1     Running   3          5d
piraeus-op-etcd-1                            1/1     Running   3          5d
piraeus-op-etcd-2                            1/1     Running   3          5d
piraeus-op-ha-controller-df776887b-j59ms     1/1     Running   150        3d22h
piraeus-op-ns-node-4whrm                     1/1     Running   1          4d18h
piraeus-op-ns-node-b4zpq                     1/1     Running   1          4d18h
piraeus-op-ns-node-pv429                     1/1     Running   1          4d18h
piraeus-op-operator-7466ddd49c-h776t         1/1     Running   4          5d
```

Here are the logs of the most recent restarts, I hope it helps:

```
time="2020-12-30T09:31:28Z" level=info msg="starting piraeus-ha-controller" version=v0.1.1                                                                                                                                                
I1230 09:31:28.799096       1 leaderelection.go:243] attempting to acquire leader lease  piraeus/piraeus-ha-controller...                                                                                                                 
I1230 09:31:28.870384       1 leaderelection.go:253] successfully acquired lease piraeus/piraeus-ha-controller                                                                                                                            
time="2020-12-30T09:31:28Z" level=info msg="new leader" leader=piraeus-op-ha-controller-df776887b-j59ms                                                                                                                                   
time="2020-12-30T09:31:28Z" level=info msg="gained leader status"                                                                                                                                                                         
time="2020-12-30T10:03:12Z" level=fatal msg="failed to run HA Controller" error="pvc updates closed unexpectedly"                                                                                                                         

time="2020-12-30T10:03:12Z" level=info msg="starting piraeus-ha-controller" version=v0.1.1                                                                                                                                                
I1230 10:03:12.539253       1 leaderelection.go:243] attempting to acquire leader lease  piraeus/piraeus-ha-controller...                                                                                                                 
I1230 10:03:12.578347       1 leaderelection.go:253] successfully acquired lease piraeus/piraeus-ha-controller                                                                                                                            
time="2020-12-30T10:03:12Z" level=info msg="gained leader status"                                                                                                                                                                         
time="2020-12-30T10:03:12Z" level=info msg="new leader" leader=piraeus-op-ha-controller-df776887b-j59ms                                                                                                                                   
time="2020-12-30T10:40:24Z" level=fatal msg="failed to run HA Controller" error="pvc updates closed unexpectedly"                                                                                                                         
 
time="2020-12-30T10:40:25Z" level=info msg="starting piraeus-ha-controller" version=v0.1.1                                                                                                                                                
I1230 10:40:25.837975       1 leaderelection.go:243] attempting to acquire leader lease  piraeus/piraeus-ha-controller...                                                                                                                 
I1230 10:40:25.870270       1 leaderelection.go:253] successfully acquired lease piraeus/piraeus-ha-controller                                                                                                                            
time="2020-12-30T10:40:25Z" level=info msg="new leader" leader=piraeus-op-ha-controller-df776887b-j59ms                                                                                                                                   
time="2020-12-30T10:40:25Z" level=info msg="gained leader status"
 ```

Related Helm values:

```yaml
haController:
  enabled: true
  image: quay.io/piraeusdatastore/piraeus-ha-controller:v0.1.1
  affinity: {}
  tolerations: []
  resources:
    limits:
      cpu: "0.2"
      memory: "250Mi"
    requests:
      cpu: "0.1"
      memory: "100Mi"
  replicas: 1
```

### PR DESCRIPTION
**Pull request by WanzenBug**
Kubernetes watches can fail after 10-30m. This caused the HA Controller
to restart _a lot_. Instead of crashing the whole app, we just restart the
watch. To do this efficiently, we keep track of certain checkpoints, so
we don't have to restart from scratch every time. While we might see some
events twice, our state of the world logic should handle this without
much trouble.

Fixes #2

### CODE REVIEW

**Comment by WanzenBug** [comment]
Just to have this documented here:
I tried using the `k8s.io/client-go/tools/cache` package for this purpose. That would automatically deal with failed watches and I would have used it directly from the start. A big problem with that package however: it can't run tests against the simple fake client. Instead we would have to spin up a fake control plane, which would make our basic tests extremely complicated. In the end I just added a simple retry loop for the watches instead.

File: pkg/hacontroller/controller.go [hunk 1]
```diff
@@ -190,7 +194,7 @@ type haController struct {
 }
 
 func (hac *haController) watchPods(ctx context.Context) (<-chan watch.Event, error) {
-	initialPods, err := hac.kubeClient.CoreV1().Pods("").List(ctx, hac.podListOpts)
+	initialPods, err := hac.kubeClient.CoreV1().Pods(metav1.NamespaceAll).List(ctx, hac.podListOpts)
```
Comment (rck) [inline pkg/hacontroller/controller.go:197]: why did that change? is it a change or the same as `""`?

Comment (WanzenBug) [reply pkg/hacontroller/controller.go:197]: Its a const `""`, so no real change

Comment (WanzenBug) [reply pkg/hacontroller/controller.go:197]: as to why: I should have used it from the start and I felt a bit silly creating a PR with 3 changed `""`

**Pull request merged**